### PR TITLE
esp8266: fix GPIO16 open drain mode

### PR DESCRIPTION
**Pull request by marcin1j**
This adds a working open drain mode for GPIO16. Looks like the original plan to overcome the lack of (apparently missing) control register for open drain driver was to switch between input and output mode with low state.
In general this solution works well but control logic for `GPIO_MODE_OPEN_DRAIN` in `set_pin() `function was missing.

There's also a minor cleanup in `mp_hal_pin_open_drain()`. Setting output register value won't hurt, but is unnecessary.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the contribution here, and sorry that this PR got a bit lost!

> There's also a minor cleanup in `mp_hal_pin_open_drain()`. Setting output register value won't hurt, but is unnecessary.

This part might actually be needed by the `mp_hal_pin_od_low(pin)` macro (would need to double check though).

File: ports/esp8266/machine_pin.c [hunk 1]
```diff
@@ -225,6 +215,16 @@ void pin_set(uint pin, int value) {
...
+    if (pin == 16) {
+        WRITE_PERI_REG(PAD_XPD_DCDC_CONF, (READ_PERI_REG(PAD_XPD_DCDC_CONF) & 0xffffffbc) | 1);
+        WRITE_PERI_REG(RTC_GPIO_CONF, READ_PERI_REG(RTC_GPIO_CONF) & ~1);
+        WRITE_PERI_REG(RTC_GPIO_ENABLE, (READ_PERI_REG(RTC_GPIO_ENABLE) & ~1) | enable);
+        if (value != -1) {
+            WRITE_PERI_REG(RTC_GPIO_OUT, (READ_PERI_REG(RTC_GPIO_OUT) & ~1) | value);
+        }
+        return;
+    }
+
```
Comment (dpgeorge) [inline ports/esp8266/machine_pin.c]: This change looks good.  How did you test it?

Comment (marcin1j) [reply ports/esp8266/machine_pin.c]: At software level by setting its value and checking that object returns exactly the same value that was set previously.
At hardware level first by a multimeter, then by actually controlling an active-low relay circuit.

**Comment by marcin1j** [comment]
> > There's also a minor cleanup in `mp_hal_pin_open_drain()`. Setting output register value won't hurt, but is unnecessary.
> 
> This part might actually be needed by the `mp_hal_pin_od_low(pin)` macro (would need to double check though).

Indeed, this needs testing as `mp_hal_pin_od_*` macros don't use `gpio_output_set()` for pin 16. Any way to call `mp_hal_pin_od_*` macros from Python code to check this?

I don't know why there's an exception for pin 16 and why call `WRITE_PERI_REG` directly. Was it for performance reasons or just `gpio_output_set()` didn't support OPEN_DRAIN for ping 16 properly? In the latter case the exception is probably not needed anymore and you can rely on `gpio_output_set()` for all pins, as esp32 port does.

**Comment by marcin1j** [comment]
> Indeed, this needs testing as `mp_hal_pin_od_*` macros don't use `gpio_output_set()` for pin 16. Any way to call `mp_hal_pin_od_*` macros from Python code to check this?

Regardless of that, it would be wise to test dht and i2c drivers using pin 16. There are three cases to test: dht, sda and scl connected to pin 16.
I can do that but I want to test `mp_hal_pin_od_*` macros alone first.

Sorry for quoting myself :)

**Comment by dpgeorge** [comment]
> Any way to call `mp_hal_pin_od_*` macros from Python code to check this?

The DHT and I2C drivers use these macros.

> I don't know why there's an exception for pin 16 and why call `WRITE_PERI_REG` directly. Was it for performance reasons

Pin 16 is different to pins 0-15.  And these macros need to be fast so that, for example, software I2C can be fast enough (eg 400kHz).

**Comment by marcin1j** [comment]
> Pin 16 is different to pins 0-15. And these macros need to be fast so that, for example, software I2C can be fast enough (eg 400kHz).
gpio_output_set() handles the differences. Surely it needs to be fast but if it's fast enough for another GPIO pins, why wouldn't it for GPIO 16?

**Comment by dpgeorge** [comment]
> gpio_output_set() handles the differences

I thought that `gpio_output_set()` only handled the GPIO0-15 pins, and GPIO 16 had to go through the RTC registers explicitly.

File: ports/esp8266/machine_pin.c (hunk 1 above)
Comment (dpgeorge) [reply ports/esp8266/machine_pin.c]: I don't think this new code will work in open-drain mode when `value==-1`: in such a case this code here won't be reached and the pin won't be configured at all.

**Comment by dpgeorge** [comment]
Closing due to inactivity.  It still needs quite a lot of testing (eg with DHT, I2C on pin 16) before it can be merged.

**Pull request closed without merging**